File: numbers/numbers.js
////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 1 - INITALIZATION
//
//  Initialize the 'newNumber' variable as any kind of number
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let newNumber = 1

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 2 - IS NUMBER
//
//  Check if variables 'isNumer1' and 'isNumber2' are numbers and save
//  the result into 'number1IsNumber' and 'number2IsNumber' respectively
let isNumber1 = "Nan"
let isNumber2 = 4000
//  NOTE: 'number1IsNumber' and 'number2IsNumber' should be boolean values
//  of True or False. The code will indeed be the same for both variables
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let number1IsNumber = !isNaN(isNumber1)
let number2IsNumber = !isNaN(isNumber2)

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 3 - ABSOLUTE VALUE
//
//  Get the absolute value of 'getMyAbsoluteValue' and save the result
//  into 'absoluteValue'
let getMyAbsoluteValue = -99
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let absoluteValue = Math.abs(getMyAbsoluteValue)

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 4 - ROUND UP A NUMBER
//
//  Round up the variable 'roundMeUp' to the nearest integer and save the
//  the result into 'roundedUpNumber'
let roundMeUp = 85.4
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let roundedUpNumber = Math.ceil(roundMeUp)

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 5 - ROUND DOWN A NUMBER
//
//  Round down the variable 'roundMeDown' to the nearest integer and save the
//  the result into 'roundedDownNumber'
let roundMeDown = 85.4
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let roundedDownNumber = Math.floor(roundMeDown)

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 6 - NUMBER TO THE POWER
//
//  Calculate 'baseNumber' to the power of 'powerNumber' and save the
//  result into 'baseToThePower'
let baseNumber = 2
let powerNumber = 9
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let baseToThePower = Math.pow(baseNumber, powerNumber)

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 7 - PERCENTAGE CALCULATION TO 2 DECIMAL PLACES
//
//  Calculate the percentage of 'scoreValue' out of 'maxValue' to
//  2 decimal places and save the result into 'scorePercentage'
let scoreValue = 67
let maxValue = 145
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let scorePercentage = ((scoreValue / maxValue) * 100).toFixed(2)

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 8 - Add number to string
//
//  Add the number in variable 'score' to the end of the string variable
//  'messagePrefix' and save it into the variable 'scoreMessage'
let messagePrefix = "You got a score of"
let score = 145
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let scoreMessage = messagePrefix + ' ' + score

////////////////////////////////////////////////////////////////////////



////////////////////////////////////////////////////////////////////////
//
//  EXERCISE 9 - Parse float from string
//
//  Parse the number from string variable 'parseTheNumber' and save it into
//  'parsedNumber' which should be of number data type
let parseTheNumber = "You got a score of 25.5"
//
////////////////////// YOUR CODE BELOW THIS LINE ///////////////////////

let parsedNumber = parseFloat(parseTheNumber.match(/[\d.]+/)[0])

////////////////////////////////////////////////////////////////////////



/////////////////////// END OF NUMBERS EXERCISES ///////////////////////



////////////////////////////////////////////////////////////////////////
//
//                               ANSWERS
//
//  DO NOT LOOK AT THIS SECTION UNTIL YOU HAVE FINISHED THE EXERCISES
//
//  The following code tests the variables you need to work on to confirm your work is
//  correct. The if conditions check the variables are correct and the console.log()
//  statements are what make the print outs you see when you run the file. The below
//  is a crude demonstration of what we refer to as code tests. Code testing is a
//  common practice in the software engineer's role, where tests are run to confirm all
//  code runs correctly. Only when all tests pass would we give ourselves the green light
//  to ship code to production and feel confident the system will perform as expected.
//
////////////////////////////////////////////////////////////////////////

// setGreen instructs the shell to set the printed output to green
let setGreen = "\x1b[32m"
// setRed instructs the shell to set the printed output to red
let setRed = "\x1b[31m"
// resetColor instructs the shell to reset the printout color to standard
let resetColor = "\x1b[0m"

// Set correctMsg and incorrectMsg Messages so I don't need to write it everytimne
let correctMsg = setGreen + "Correct! " + resetColor
let incorrectMsg = setRed + "Incorrect! " + resetColor

// totalTests and passedTests count the final tally of passed tests
let totalTests = 0
let passedTests = 0

// Number Initialization
if (newNumber !== undefined && typeof newNumber === "number") {
  console.log(correctMsg + "You initialized the 'newNumber' number variable as " + newNumber)
  passedTests++
} else {
  console.log(incorrectMsg + "The initialized 'newNumber' should be defined and be a number, but yours is " + newNumber)
}
totalTests++

// Variable Is Number 1
let correctNumber1IsNumber = typeof isNumber1 === "number"
if (number1IsNumber !== undefined && number1IsNumber == correctNumber1IsNumber) {
  console.log(correctMsg + "As for the variable 'isNumber1' being a number, you found this was " + number1IsNumber)
  passedTests++
} else {
  console.log(incorrectMsg + "As for the variable 'isNumber1' being a number, this should be " + correctNumber1IsNumber + " but you got " + number1IsNumber)
}
totalTests++

// Variable Is Number 2
let correctNumber2IsNumber = typeof isNumber2 === "number"
if (number2IsNumber !== undefined && number2IsNumber == correctNumber2IsNumber) {
  console.log(correctMsg + "As for the variable 'isNumber1' being a number, you found this was " + number2IsNumber)
  passedTests++
} else {
  console.log(incorrectMsg + "As for the variable 'isNumber2' being a number, this should be " + correctNumber2IsNumber + " but you got " + number2IsNumber)
}
totalTests++

// Absolute Value
let correctAbsoluteValue = Math.abs(getMyAbsoluteValue)
if (absoluteValue !== undefined && absoluteValue == correctAbsoluteValue) {
  console.log(correctMsg + "The absolute value of 'getMyAbsoluteValue' you got was " + absoluteValue)
  passedTests++
} else {
  console.log(incorrectMsg + "The absolute value of 'getMyAbsoluteValue' should be " + correctAbsoluteValue + " but you got " + absoluteValue)
}
totalTests++

// Round Up A Number
let correctRoundedUpNumber = Math.ceil(roundMeUp)
if (roundedUpNumber !== undefined && roundedUpNumber === correctRoundedUpNumber) {
  console.log(correctMsg + "The rounded up value of 'roundMeUp' you got was " + roundedUpNumber)
  passedTests++
} else {
  console.log(incorrectMsg + "The rounded up value of 'roundMeUp' should be " + correctRoundedUpNumber + " but you got " + roundedUpNumber)
}
totalTests++

// Round Down A Number
let correctRoundedDownNumber = Math.floor(roundMeDown)
if (roundedDownNumber !== undefined && roundedDownNumber === correctRoundedDownNumber) {
  console.log(correctMsg + "The rounded down value of 'roundMeDown' you got was " + roundedDownNumber)
  passedTests++
} else {
  console.log(incorrectMsg + "The rounded down value of 'roundMeDown' should be " + correctRoundedDownNumber + " but you got " + roundedDownNumber)
}
totalTests++

// Number to the power
let correctBaseToThePower = baseNumber**powerNumber
if (baseToThePower !== undefined && baseToThePower === correctBaseToThePower) {
  console.log(correctMsg + "The 'baseNumber' to the power of 'powerNumber' you got was " + baseToThePower)
  passedTests++
} else {
  console.log(incorrectMsg + "The 'baseNumber' to the power of 'powerNumber' should be " + correctBaseToThePower + " but you got " + baseToThePower)
}
totalTests++

// Percentage Calculation To 2 Decimal Places
let correctScorePercentage = ((scoreValue / maxValue) * 100).toFixed(2)
if (scorePercentage !== undefined && scorePercentage === correctScorePercentage) {
  console.log(correctMsg + "The percentage of 'scoreValue' out of 'maxValue' to 2 decimal places you got was " + scorePercentage)
  passedTests++
} else {
  console.log(incorrectMsg + "The percentage of 'scoreValue' out of 'maxValue' to 2 decimal places should be " + correctScorePercentage + " but you got " + scorePercentage)
}
totalTests++

// Add number to string
let correctScoreMessage = messagePrefix + " " + score
if (scoreMessage !== undefined && scoreMessage === correctScoreMessage) {
  console.log(correctMsg + "The vale of 'scoreMessage' you got was " + scoreMessage)
  passedTests++
} else {
  console.log(incorrectMsg + "The vale of 'scoreMessage' should be '" + correctScoreMessage + "' but you got " + scoreMessage)
}
totalTests++

// Parse number from string
let correctParsedNumber = parseFloat(parseTheNumber.replace(/[^0-9.]/g, ""))
if (parsedNumber !== undefined && parsedNumber === correctParsedNumber) {
  console.log(correctMsg + "The vale of 'parsedNumber' you got was " + parsedNumber + " of type " + typeof(parsedNumber))
  passedTests++
} else {
  console.log(incorrectMsg + "The vale of 'parsedNumber' should be " + correctParsedNumber + " of type " + typeof(correctParsedNumber) + " but you got " + parsedNumber + " of type " + typeof(parsedNumber))
}
totalTests++

// Final Test Tally Determination
if (passedTests === totalTests) {
  console.log(setGreen + "You passed " + passedTests + " / " + totalTests + " tests. Congratulations!" + resetColor)
} else {
  console.log(setRed + "You passed " + passedTests + " / " + totalTests + " tests. Keep Going!" + resetColor)
}

File: numbers/numbers.py
########################################################################
#
#  EXERCISE 1 - INTEGER INITALIZATION
#
#  Initialize the 'newInteger' variable as any kind of INTEGER
#
###################### YOUR CODE BELOW THIS LINE #######################

newInteger = 1

########################################################################



########################################################################
#
#  EXERCISE 2 - FLOAT INITALIZATION
#
#  Initialize the 'newFloat' variable as any kind of FLOAT
#
###################### YOUR CODE BELOW THIS LINE #######################

newFloat = 3.2

########################################################################



########################################################################
#
#  EXERCISE 3 - IS INTEGER
#
#  Check if variable 'integer1' is an integer and save the result into
#  'integer1IsInteger'
integer1 = "Not an integer"
#  NOTE: integer1IsInteger should be a boolean value of True or False
#
###################### YOUR CODE BELOW THIS LINE #######################

integer1IsInteger = isinstance(integer1, int)

########################################################################



########################################################################
#
#  EXERCISE 4 - IS FLOAT
#
#  Check if variable 'float1' is a float and save the result into
#  'float1IsFloat'
float1 = 45.6
#  NOTE: float1IsFloat should be a boolean value of True or False
#
###################### YOUR CODE BELOW THIS LINE #######################

float1IsFloat = isinstance(float1, float)

########################################################################



########################################################################
#
#  EXERCISE 5 - ABSOLUTE VALUE
#
#  Get the absolute value of 'getMyAbsoluteValue' and save the result
#  into 'absoluteValue'
getMyAbsoluteValue = -64.4
#
###################### YOUR CODE BELOW THIS LINE #######################

absoluteValue = abs(getMyAbsoluteValue)

########################################################################



########################################################################
#
#  EXERCISE 6 - ROUND UP A NUMBER
#
#  Round up the variable 'roundMeUp' to the nearest integer and save the
#  the result into 'roundedUpNumber'
roundMeUp = 85.4
#  NOTE: You will need to use the Mathematics module 'math' imported
#  below for you
import math
#
###################### YOUR CODE BELOW THIS LINE #######################

roundedUpNumber = math.ceil(roundMeUp)

########################################################################



########################################################################
#
#  EXERCISE 7 - ROUND DOWN A NUMBER
#
#  Round down the variable 'roundMeDown' to the nearest integer and save the
#  the result into 'roundedDownNumber'
roundMeDown = 85.4
#  NOTE: You will need to use the Mathematics module 'math' imported
#  in the previous exercise for you
#
###################### YOUR CODE BELOW THIS LINE #######################

roundedDownNumber = math.floor(roundMeDown)

########################################################################



########################################################################
#
#  EXERCISE 8 - NUMBER TO THE POWER
#
#  Calculate 'baseNumber' to the power of 'powerNumber' and save the
#  result into 'baseToThePower'
baseNumber = 4
powerNumber = 5
#
###################### YOUR CODE BELOW THIS LINE #######################

baseToThePower = math.pow(4, 5)

########################################################################



########################################################################
#
#  EXERCISE 9 - PERCENTAGE CALCULATION TO 3 DECIMAL PLACES
#
#  Calculate the percentage of 'scoreValue' out of 'maxValue' to
#  3 decimal places and save the result into 'scorePercentage'
scoreValue = 23.5
maxValue = 35
#
###################### YOUR CODE BELOW THIS LINE #######################

unrounded = (scoreValue / maxValue) * 100
scorePercentage = round(unrounded, 3)
########################################################################



########################################################################
#
#  EXERCISE 10 - Add number to string
#
#  Add the number in variable 'score' to the end of the string variable
#  'messagePrefix' and save it into the variable 'scoreMessage'
messagePrefix = "You got a score of"
score = 145
#
###################### YOUR CODE BELOW THIS LINE ######################

scoreMessage = messagePrefix + ' '+ str(score)

########################################################################



########################################################################
#
#  EXERCISE 9 - Parse float from string
#
#  Parse the number from string variable 'parseTheNumber' and save it into
#  'parsedNumber' which should be of number data type
parseTheNumber = "Prerequisite score of 50.5 is required to pass"
#  NOTE: You will need to use the imported regular expression module 're'
#  imported below for you
import re
#
###################### YOUR CODE BELOW THIS LINE ######################

match = re.search(r'\d+\.?\d*', parseTheNumber)
parsedNumber = float(match.group())

########################################################################



####################### END OF STRINGS EXERCISES #######################



########################################################################
#
#								ANSWERS
#
#  DO NOT LOOK AT THIS SECTION UNTIL YOU HAVE FINISHED THE EXERCISES
#
#  The following code tests the variables you need to work on to confirm your work is
#  correct. The if conditions check the variables are correct and the console.log()
#  statements are what make the print outs you see when you run the file. The below
#  is a crude demonstration of what we refer to as code tests. Code testing is a
#  common practice in the software engineer's role, where tests are run to confirm all
#  code runs correctly. Only when all tests pass would we give ourselves the green light
#  to ship code to production and feel confident the system will perform as expected.
#
########################################################################

# setGreen instructs the shell to set the printed output to green
setGreen = "\x1b[32m"
# setRed instructs the shell to set the printed output to red
setRed = "\x1b[31m"
# resetColor instructs the shell to reset the printout color to standard
resetColor = "\x1b[0m"

# Set correctMsg and incorrectMsg Messages so I don't need to write it everytimne
correctMsg = setGreen + "Correct! " + resetColor
incorrectMsg = setRed + "Incorrect! " + resetColor

# totalTests and passedTests count the final tally of passed tests
totalTests = 0
passedTests = 0

# Integer Initialization
if newInteger == None:
	print(incorrectMsg + "The initialized 'newInteger' should be defined")
elif type(newInteger) is int:
	print(correctMsg + "You initialized 'newInteger' as: '" + str(newInteger))
	passedTests+= 1
else:
	print(incorrectMsg + "The initialized 'newInteger' should be an integer, yours is " +	str(type(newInteger)))
totalTests += 1

# Float Initialization
if newFloat == None:
	print(incorrectMsg + "The initialized 'newFloat' should be defined")
elif type(newFloat) is float:
	print(correctMsg + "You initialized 'newFloat' as '" + str(newFloat))
	passedTests+= 1
else:
	print(incorrectMsg + "The initialized 'newFloat' should be a float, yours is " +	str(type(newFloat)))
totalTests += 1

# Is Integer
correctInteger1IsInteger = type(integer1) is int
if integer1IsInteger == None:
	print(incorrectMsg + "The 'integer1IsInteger' variable should be defined")
elif integer1IsInteger == correctInteger1IsInteger:
	print(correctMsg + "As for the variable 'integer1' being an integer, you found this was " + str(integer1IsInteger))
	passedTests+= 1
else:
	print(incorrectMsg + "As for the variable 'integer1' being an integer, this should be " + str(correctInteger1IsInteger) + " but you got " + str(integer1IsInteger))
totalTests += 1

# Is Float
correctFloat1IsFloat = type(float1) is float
if float1IsFloat == None:
	print(incorrectMsg + "The 'float1IsFloat' variable should be defined")
elif float1IsFloat == correctFloat1IsFloat:
	print(correctMsg + "As for the variable 'float1' being a float, you found this was " + str(float1IsFloat))
	passedTests+= 1
else:
	print(incorrectMsg + "As for the variable 'float1' being a float, this should be " + str(correctFloat1IsFloat) + " but you got " + str(float1IsFloat))
totalTests += 1

# Absolute Value
correctAbsoluteValue = abs(getMyAbsoluteValue)
if absoluteValue == None:
	print(incorrectMsg + "The 'absoluteValue' variable should be defined")
elif absoluteValue == correctAbsoluteValue:
	print(correctMsg + "The absolute value of 'getMyAbsoluteValue' you got was " + str(absoluteValue))
	passedTests+= 1
else:
	print(incorrectMsg + "The absolute value of 'getMyAbsoluteValue' should be " + str(correctAbsoluteValue) + " but you got " + str(absoluteValue))
totalTests += 1

# Round Up A Number
correctRoundedUpNumber = math.ceil(roundMeUp)
if roundedUpNumber == None:
	print(incorrectMsg + "The 'roundedUpNumber' variable should be defined")
elif roundedUpNumber == correctRoundedUpNumber:
	print(correctMsg + "The rounded up number of 'roundMeUp' you got was " + str(roundedUpNumber))
	passedTests+= 1
else:
	print(incorrectMsg + "The rounded up number of 'roundMeUp' should be " + str(correctRoundedUpNumber) + " but you got " + str(roundedUpNumber))
totalTests += 1

# Round Down A Number
correctRoundedDownNumber = math.floor(roundMeDown)
if roundedDownNumber == None:
	print(incorrectMsg + "The 'roundedDownNumber' variable should be defined")
elif roundedDownNumber == correctRoundedDownNumber:
	print(correctMsg + "The rounded down number of 'roundMeDown' you got was " + str(roundedDownNumber))
	passedTests+= 1
else:
	print(incorrectMsg + "The rounded down number of 'roundMeDown' should be " + str(correctRoundedDownNumber) + " but you got " + str(roundedDownNumber))
totalTests += 1

# Number To The Power
correctBaseToThePower = baseNumber**powerNumber
if baseToThePower == None:
	print(incorrectMsg + "The 'roundedDownNumber' variable should be defined")
elif baseToThePower == correctBaseToThePower:
	print(correctMsg + "The 'baseNumber' to the power of 'powerNumber' you got was " + str(baseToThePower))
	passedTests+= 1
else:
	print(incorrectMsg + "The 'baseNumber' to the power of 'powerNumber' should be " + str(correctBaseToThePower) + " but you got " + str(baseToThePower))
totalTests += 1

# Percentage Calculation To 3 Decimal Places
correctScorePercentage = round((100 * scoreValue / maxValue), 3)
if scorePercentage == None:
	print(incorrectMsg + "The 'scorePercentage' variable should be defined")
elif scorePercentage == correctScorePercentage:
	print(correctMsg + "The percentage of 'scoreValue' out of 'maxValue' to 3 decimal places you got was " + str(scorePercentage))
	passedTests+= 1
else:
	print(incorrectMsg + "The percentage of 'scoreValue' out of 'maxValue' to 3 decimal places should be " + str(correctScorePercentage) + " but you got " + str(scorePercentage))
totalTests += 1

# Add number to string
correctScoreMessage =	messagePrefix + " " + str(score)
if scoreMessage == None:
	print(incorrectMsg + "The 'scoreMessage' variable should be defined")
elif scoreMessage == correctScoreMessage:
	print(correctMsg + "The value of 'scoreMessage' you got was " + scoreMessage)
	passedTests+= 1
else:
	print(incorrectMsg + "The value of 'scoreMessage' should be " + correctScoreMessage + " but you got " + scoreMessage)
totalTests += 1

# Parse number from string
correctParsedNumber = float(re.sub(r'[^0-9.]', "", parseTheNumber))
if parsedNumber == None:
	print(incorrectMsg + "The 'parsedNumber' variable should be defined")
elif parsedNumber == correctParsedNumber:
	print(correctMsg + "The value of 'parsedNumber' you got was " + str(parsedNumber))
	passedTests+= 1
else:
	print(incorrectMsg + "The value of 'parsedNumber' should be " + str(correctParsedNumber) + " of type " + str(type(correctParsedNumber)) + " but you got " + str(parsedNumber)	+ " of type " + str(type(parsedNumber)))
totalTests += 1

# Final Test Tally Determination
if passedTests == totalTests:
	print(setGreen + "You passed " + str(passedTests) + " / " + str(totalTests) + " tests. Congratulations!" + resetColor)
else:
	print(setRed + "You passed " + str(passedTests) + " / " + str(totalTests) + " tests. Keep Going!" + resetColor)
